cpu/sam3x8e/periph/pwm.c

```diff
@@ -0,0 +1,258 @@
+
+    Pwm* pwm;
+    PwmCh_num* chan;
+
+    /*
+     * Mode check.
+     * Involves things that shold be in "periph_conf.h".
+     * !Junk!Just wasting memory an a few cpu cycles for now. ;/
+     */
+    switch ( mode) {
+    case PWM_LEFT:
+        //!\todo
+        break;
+    case PWM_CENTER:
+        //!\todo
+        break;
+    default:
+        //retval = ERR_INIT_BWTH;
+        // Don't like :/
+        return ERR_INIT_BWTH;
+    }
+
+    /* Should check if "|log_2 frequency|+|log_2 resolution| <= 32" */
+    pwm_clk = frequency * resolution;
+
+    /*
+     * Sanity check.
+     *
+     * The pwm provides 11 prescaled clocks with (MCK/2^prea | prea=[0,10])
+     * and a divider (diva) with a denominator range [1,255] in line.
+     * With every step the prescaler is doubled and the remaining quotient
+     * is halfed and tested <=255.
+     */
+    if ( F_CPU < pwm_clk) { // Have to cut down resulting frequency.
+        frequency = F_CPU/resolution;
+    } else { // Estimate prescaler and divider.
+        diva = F_CPU / pwm_clk;
+        while ( ( prea < MCK_DIV_LB_MAX) && ( ~0xff & diva)) {
+            prea = prea + 1;
+            diva = diva >> 1;
+        }
+        frequency = F_CPU / ( ( resolution * diva) << prea);
+    }
+
+    retval = frequency;
+
+    /* Activate PWM block by enabling it's clock. */
+    PMC->PMC_PCER1 = PMC_PCER1_PID36;
+
+    /*!
+     * PWM ID is ID_PWM (36).
+     * \note { Is there any choice? }
+     */
+    pwm = PWM_0_DEV;
+
+    /*
+     * ++ Initialisation ++
+     * adviced by "38.6.5.1 Initialization" in
+     * "Atmel SAM3X8E  SAM3X8C SAM3X4E SAM3X4C SAM3A8C SAM3A8C Datasheet"
+     */
+    /* Step 1. Unlock User Interface */
+    pwm->PWM_WPCR  = PWM_ENA_CHID0;
+
+    /* Step 2. Configure clock generator */
+    pwm->PWM_CLK = ( 0xffff0000 & pwm->PWM_CLK)|
+                   // ( prea << 8) | diva | 0x1;
+                   PWM_CLK_PREA(prea)|PWM_CLK_DIVA(diva);
+
+    /* +++++++++++ Create and setup sync group +++++++++++++++ */
+    /*!
+     * Work dedicated on pwm channel N.
+     */
+    chan = &(pwm->PWM_CH_NUM[0]);
+    /* Step 3. Select clock + mode */
+    chan->PWM_CMR = PWM_CMR_CPRE_CLKA;
+    /* Step 7. Configuration of the period for CH(N) */
+    chan->PWM_CPRD = resolution - 1;
+    /* Step 8. Configuration of the duty-cycle for CH(N) */
+    chan->PWM_CDTY = 0;
+
+    /*!
+     * Step 10. Select sync'ed channels (4-7)
+     *
+     * \todo { replace literals with CMSIS expressions }
+     */
+    pwm->PWM_SCM = ( pwm->PWM_SCM & 0xfffcff00) | ( 0x1 << 16 ) |
+                   PWM_SCM_SYNC4 | PWM_SCM_SYNC5 | PWM_SCM_SYNC6 | PWM_SCM_SYNC7;
+
+    /* Step 19. Enable the PWM channel(s). */
+    pwm->PWM_ENA = PWM_ENA_CHID0;
+
+    /* Protect registers, later */
+    //pwm->PWM_WPCR  = PWM_ENA_CHID0;
+
+    /* +++++++++++++ Sync group ready +++++++++++++++++++++++++ */
+
+
+    /* Connect channel outputs to PINs */
+    PWM_0_CHAN_PIN_SET();
```

Comment:
You should be absolutely independent from the pwm-channel or pin here. The driver should work for any configuration. That's why I proposed to introduce a function or just place the regarding code here.